cpp/src/arrow/array/builder_union.h

```diff
@@ -155,6 +156,21 @@ class ARROW_EXPORT DenseUnionBuilder : public BasicUnionBuilder {
     return offsets_builder_.Append(offset);
   }
 
+  Status AppendArraySliceUnchecked(const ArrayData& array, const int64_t offset,
+                                   const int64_t length) override {
```

Comment:
The definition can probably moved to `builder_union.cc` instead?
(same for SparseUnionBuilder)